src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -47,61 +57,204 @@ private void processInput(String userInput) {
      */
     public final void toggleExactMatchMilestone(String milestoneName) {
-        Optional<PickerMilestone> onlyMatchingMilestone = getExactMatchMilestone(currentMilestonesList, milestoneName);
-        onlyMatchingMilestone.ifPresent(milestone -> toggleMilestone(milestone));
+        Optional<PickerMilestone> exactMatchMilestone = getExactMatchMilestone(currentMilestones, milestoneName);
+        exactMatchMilestone.ifPresent(this::toggleMilestone);
     }
 
     /**
-     * Toggles the milestone that is the first match to the given query
-     * <p>
-     * This method is NOT case-sensitive
+     * Toggles the first milestone of the list
      *
      * @param milestoneList
      */
-    private final void toggleFirstMatchingMilestone(List<PickerMilestone> milestoneList) {
-        Optional<PickerMilestone> firstMatchingMilestone = getFirstMatchingMilestone(milestoneList);
+    private final void toggleFirstMilestone(List<PickerMilestone> milestoneList) {
+        Optional<PickerMilestone> firstMatchingMilestone = milestoneList.stream()
+                .findFirst();
         firstMatchingMilestone.ifPresent(this::toggleMilestone);
     }
 
     /**
-     * Changes the selection statuses of all milestones in the milestone list,
-     * such that the new given milestone is toggled, and the rest are not selected
+     * Changes the selection statuses of all milestones in the currentMilestones,
+     * such that the given milestone is toggled, and the rest are not selected
      *
      * @param milestone
      */
     private void toggleMilestone(PickerMilestone milestone) {
-        this.currentMilestonesList.stream()
-                .forEach(listMilestone ->
-                        listMilestone.setSelected(listMilestone.equals(milestone) && !listMilestone.isSelected()));
+        this.currentMilestones.stream()
+                .forEach(listMilestone -> listMilestone.setSelected(listMilestone.equals(milestone)
+                        && !listMilestone.isSelected()));
+    }
+
+    public List<PickerMilestone> getCurrentMilestones() {
+        return currentMilestones;
+    }
+
+    public List<PickerMilestone> getBestMatchingMilestones() {
+        return bestMatchingMilestones;
+    }
+
+    /**
+     * Generates the best matching milestones which will then be added to bestMatchingMilestones
+     *
+     * The added milestones are references to the actual elements in currentMilestones, and thus should not be
+     * unnecessarily mutated
+     * @param userInput
+     */
+    private void determineBestMatchingMilestones(String userInput) {
+        addMatchingMilestones(bestMatchingMilestones, getNumberOfMilestonesToAdd(), currentMilestones);
+        addPreviouslyMatchingMilestones(bestMatchingMilestones, getNumberOfMilestonesToAdd(), userInput,
+                                        currentMilestones);
+        addLikelyUnmatchedMilestones(bestMatchingMilestones, getNumberOfMilestonesToAdd(), currentMilestones);
```

Comment:
This was the initial implementation but it felt a bit overwhelming for a top-level method (with all the braces, variable names, etc). Will change it back.